### PR TITLE
Remove rechecking/refreshed rate limit functionality

### PR DESCRIPTION
**Pull request by bitwiseman**
The ApiRateLimitChecker included functionality to handle when the rate limit is reset "before expected".
If a real world scenario exists for this behavior, it is an extreme edge case. Meanwhile, the functionality
add unneeded complexity to the checkers and is not supported by the github-api RateLimitChecker class.

This change remove this extraneous functionality and fixes the test match.

This change in need to move to using the built-in github-api `RateLimitChecker` functionality.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java [hunk 1]
```diff
@@ -227,40 +248,32 @@ protected boolean checkRateLimit(GHRateLimit.Record rateLimitRecord,
...
-                            writeLog("Jenkins-Imposed API Limiter: The Github API usage quota may have been refreshed earlier than expected, rechecking...");
-                            break;
-                        }
-                    } catch (Exception e) {
-                        // This code path is a contigency only.  If it doesn't work, just continue sleeping
-                    }
-                    writeLog(String.format(
-                        "Jenkins-Imposed API Limiter: Still sleeping, now only %s remaining.",
-                        Util.getTimeSpanString(expiration - now)
-                    ));
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:262]: This is the removed functionality.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java [hunk 2]
```diff
@@ -241,29 +243,37 @@ public void ThrottleOnOverTest() throws Exception {
...
         //should be no output
         assertEquals(0, countOfOutputLines(m -> m.matches(".*[sS]leeping.*")));
 
         assertEquals(11, getRequestCount(githubApi));
 
-        // check rate limit to hit the next 10 scenarios
-        ApiRateLimitChecker.ThrottleOnOver.checkApiRateLimit(listener, github);
+        // check rate limit to hit the next 9 scenarios
+        for (int i = 0; i < 10; i++) {
+            assertTrue(ApiRateLimitChecker.ThrottleOnOver.checkApiRateLimitOnce(listener, github, i));
```
Comment (bitwiseman) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java:256]: By removing "rechecking" we remove the ability of tests to shortcut long wait loops by providing updated data. 
Instead we not do these waits manually for only a specified number of iterations in testing.

**Comment by bitwiseman** [comment]
In the next change, I'll replace the `getRateLimit()` calls with `rateLimit()` calls which will basically match the `github-api` `RateLimitChecker` behavior.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java [hunk 3]
```diff
@@ -280,6 +290,11 @@ public void ThrottleForNormalizeTestWithinIdeal() throws Exception {
         // Approximate the ideal here
         int approximateIdeal = 4000;
 
+        // NOTE: The behavior below is not longer interesting.
```
Comment (timja) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java:293]:
```suggestion
        // NOTE: The behavior below is no longer interesting.
```

**Review by timja — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java [hunk 4]
```diff
@@ -47,29 +48,8 @@ boolean checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count) th
...
-                            } else {
-                                expiration = rateLimit.getResetDate().getTime() + ENTROPY.nextInt(
-                                    EXPIRATION_WAIT_MILLIS);
-                                writeLog(String.format(
-                                            "Jenkins-Imposed API Limiter: Current quota for Github API usage has %d remaining (%d over budget). Next quota of %d in %s. Sleeping until reset.",
-                                            rateLimit.getRemaining(),
-                                            ideal - rateLimit.getRemaining(),
-                                            rateLimit.getLimit(),
-                                            Util.getTimeSpanString(rateLimitResetMillis)));
-                            }
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:72]: Move to a function for reuse.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java [hunk 5]
```diff
@@ -207,61 +178,93 @@ public String getDisplayName() {
...
+            }
             return this.checkRateLimitImpl(rateLimitRecord, count);
         }
 
+        // internal for testing
         abstract boolean checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count) throws InterruptedException;
 
-        protected void waitUntilRateLimit(GHRateLimit.Record rateLimit, long expiration, long count) throws InterruptedException {
-            long now = System.currentTimeMillis();
+        static long calculateExpirationForMinimum(GHRateLimit.Record rateLimit, long now, long rateLimitResetMillis, int minimum) {
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:219]: the name is a bit confusing to me, minimum what, minimum rateLimit doesn't seem right to me if you include parameters when reading?

calculate expiration would possibly read better

**Review by timja — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java (hunk 5 above)
Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:219]: `calculateExpirationIfOverBudgetIntoBuffer`

Comment (timja) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:219]: what about calculateExpiration, and put the quirks in the javadoc?

Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java]: It can't be `calculateExpiration` because it is only for the case where we're over budget into the buffer zone. 

Went with `calculateExpirationWhenBufferExceeded`
